### PR TITLE
Consolidate variant build steps

### PR DESCRIPTION
**Pull request by scruplelesswizard**
Creates an action with common build steps across the following variants:
- ESP32
- ESP32-C3
- ESP32-S3
- NRF52
- Raspberry Pi 2040
- STM32

### CODE REVIEW

**Review by fifieldt — APPROVED** [review]
Extra maintainable, nice!

File: .github/actions/build-variant/action.yml [hunk 1]
```diff
@@ -0,0 +1,97 @@
...
+runs:
+  using: composite
+  steps:
+    - uses: actions/checkout@v4
+
+    - name: Build base
+      id: base
+      uses: ./.github/actions/setup-base
+
+    - name: Pull web ui
```
Comment (fifieldt) [inline .github/actions/build-variant/action.yml:44]: In one of the tests just now "Pull web ui" failed, but there was no log explaining as to why. Just wondering, is there any easy config you can add here that would give a bit more debug info? If not easy, please ignore this comment :)

**Comment by fifieldt** [comment]
@scruplelesswizard - this results in errors on run:

Error: Can't find 'action.yml', 'action.yaml' or 'Dockerfile' under '/home/ubuntu/runner/_work/firmware/firmware/.github/actions/build-variant'. Did you forget to run actions/checkout before running your local action?

**Comment by scruplelesswizard** [comment]
> Did you forget to run actions/checkout before running your local action?

I absolutely did! 😂

File: .github/actions/build-variant/action.yml (hunk 1 above)
Comment (scruplelesswizard) [reply .github/actions/build-variant/action.yml:44]: Looks like a rerun resolved it. That one was weird, there were no logs for the one run, but there were for other runs

**Pull request merged**